scalafix-rules/src/main/scala/scalafix/internal/pc/Embedded.scala

```diff
@@ -0,0 +1,112 @@
+package scalafix.internal.pc
+
+import java.net.URLClassLoader
+import java.util.ServiceLoader
+
+import scala.jdk.CollectionConverters._
+
+import scala.meta.pc.PresentationCompiler
+
+import coursierapi.Dependency
+import coursierapi.Fetch
+import coursierapi.MavenRepository
+import scala.collection.concurrent.TrieMap
+
+object Embedded {
+
+  private val presentationCompilers: TrieMap[String, URLClassLoader] =
+    TrieMap.empty
+
+  def presentationCompiler(
+      scalaVersion: String
+  ): PresentationCompiler = {
+    val classloader = presentationCompilers.getOrElseUpdate(
+      scalaVersion,
+      newPresentationCompilerClassLoader(scalaVersion)
+    )
+    val presentationCompilerClassname =
+      if (supportPresentationCompilerInDotty(scalaVersion)) {
+        "dotty.tools.pc.ScalaPresentationCompiler"
+      } else {
+        "scala.meta.pc.ScalaPresentationCompiler"
+      }
+
+    serviceLoader(
+      classOf[PresentationCompiler],
+      presentationCompilerClassname,
+      classloader
+    )
+  }
+
+  private def supportPresentationCompilerInDotty(scalaVersion: String) = {
+    scalaVersion
+      .replaceAll(raw"-RC\d+", "")
+      .split("\\.")
+      .take(3)
+      .map(_.toInt) match {
+      case Array(3, minor, patch) => minor > 3 || minor == 3 && patch >= 4
+      case _ => false
+    }
+  }
+
+  private def scala3PresentationCompilerDependencies(version: String) =
+    if (supportPresentationCompilerInDotty(version)) {
+      val dep = Dependency
+        .of("org.scala-lang", "scala3-presentation-compiler_3", version)
+
+      // some versions of the presentation compiler depend on versions only build for JDK 11
+      dep.addExclusion("org.eclipse.lsp4j", "org.eclipse.lsp4j")
+      dep.addExclusion("org.eclipse.lsp4j", "org.eclipse.lsp4j.jsonrpc")
+      // last built with JDK 8
+      val lsp4jDep =
+        Dependency.of("org.eclipse.lsp4j", "org.eclipse.lsp4j", "0.20.1")
+      List(dep, lsp4jDep)
```

Comment:
In build.sbt, these exclusions are applied to mtags and mtags only. Maybe we don't test properly? 🤔

Reply:
Not sure if I understand, I added those exclusions there as well. It was failing previously with those

Reply:
Here we exclude the lsp4j that the compiler brings transitively, while in build.sbt, we exclude the lsp4j that mtags-interfaces brings transitively. I did not check if lsp4j was actually a dependency from one or another (or both?), but I just find this suspicious to have different logic for static vs dynamic resolution.

Maybe coursier does the exclusion for all artifacts involved for a single resolution so it's equivalent? The API just seems to show that exclusion is selective.

Reply:
I am removing mtags and direct references to mtags-interfaces for now